[WebInspector] Display OpenGL object ids in canvas inspector

djg opened the pull request:
#### 8574b90773edf4ffb2bd7f9c5fd1b361cef52360
<pre>
[WebInspector] Display OpenGL object ids in canvas inspector
<a href="https://bugs.webkit.org/show_bug.cgi?id=246186">https://bugs.webkit.org/show_bug.cgi?id=246186</a>
rdar://problem/100875691
Reviewed by Devin Rousso and Dean Jackson.

Record the underlying OpenGL object id associated with each WebGL object and
display these ids along with the WebGL object type in WebInspector canvas
inspector mode to aid determining which objects are being referred to in the
call trace.

* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl-full-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl-snapshots-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl2-full-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl2-snapshots-expected.txt:
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::processArgument):
* Source/WebInspectorUI/UserInterface/Views/RecordingActionTreeElement.js:
(WI.RecordingActionTreeElement._generateDOM.createParameterElement):
(WI.RecordingActionTreeElement._generateDOM):

Canonical link: <a href="https://commits.webkit.org/255565@main">https://commits.webkit.org/255565@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/694a296ee8cf0bd31baa6f513b75f8d9f73fdede

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92650 "24 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1865 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23231 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102367 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96651 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1865 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30216 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85035 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98528 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98314 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1255 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79140 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28177 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83171 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82881 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71267 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36624 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16803 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34415 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17986 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3852 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38286 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40592 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40198 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37146 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/inspector/InspectorCanvas.cpp:
```diff
@@ -610,35 +610,35 @@ std::optional<InspectorCanvasCallTracer::ProcessedArgument> InspectorCanvas::pro
     if (!argument)
         return std::nullopt;
-    return {{ JSON::Value::create(0), RecordingSwizzleType::WebGLBuffer }};
+    return {{ JSON::Value::create(static_cast<int>(argument->object())), RecordingSwizzleType::WebGLBuffer }};
```
Style checker complains about the `{{` but I kept the existing style instead of reformatting.

djg commented on Source/WebInspectorUI/UserInterface/Views/RecordingActionTreeElement.js:
```diff
@@ -101,6 +105,7 @@ WI.RecordingActionTreeElement = class RecordingActionTreeElement extends WI.Gene
             case WI.Recording.Swizzle.WebGLVertexArrayObject:
                 parameterElement.classList.add("swizzled");
                 parameterElement.textContent = WI.Recording.displayNameForSwizzleType(swizzleType);
+                parameterElement.textContent += ": " + parameter;
```
If someone wants to style this differently, I'd welcome it. I'm not a UI designer.

djg replied:
I like to draw something like the little colored letters in this image. 
![image](https://webkit.org/wp-content/uploads/Retained-Size.png)

dcrousso commented on Source/WebCore/inspector/InspectorCanvas.cpp:
```diff
@@ -610,35 +610,35 @@ std::optional<InspectorCanvasCallTracer::ProcessedArgument> InspectorCanvas::pro
     if (!argument)
         return std::nullopt;
-    return {{ JSON::Value::create(0), RecordingSwizzleType::WebGLFramebuffer }};
+    return {{ JSON::Value::create(static_cast<int>(argument->object())), RecordingSwizzleType::WebGLFramebuffer }};
```
Is this number something that is usable by other JS APIs?  Or is it purely internal state?

If the latter, I don't think we want this to be something we expose as there's not really a benefit to the developer.  If your goal is to just have some sort of way to differentiate between different instances of the same object (e.g. "Framebuffer #1" vs "Framebuffer #2"), then I'd argue for having a `HashMap` of identifiers instead of using the underlying `GCGLuint`.

 It's also probably gonna give you grief when dumping the recording JSON if these numbers are not deterministic.

grorg replied:
Generally they are not usable, but I think it's ok to expose them.

dcrousso replied:
maybe we should have it match how we show internal identifiers for heap snapshots? <https://webkit.org/blog/6425/memory-debugging-with-web-inspector/>

djg replied:
My goal is to have a way to work out which WebGL object is being referred to by WebGL calls in the trace. Currently, I see `glUseProgram(WebGLProgram)` which is completely useless when using the Web Inspector to try to diagnose rendering issues. (Similarly, there's no buffer data, so I don't know what data is being passed to draw calls)

OpenGL names all objects via integer ids. They're completely free-form, but there are OpenGL calls that generate them. OpenGL developers are used to using these ids to connect calls in traces.

What's the difference between displaying "Framebuffer #1" and "Framebuffer 15"? How would I store the `HashMap` of identifiers in the `InspectorCanvasCallTracer`?

dcrousso replied:
I absolutely agree that we should display something for these.

The problem with this approach, however, is that it doesn't really help the developer understand "which buffer is this".  Just because WebKit happens to call this "Framebuffer #2" doesn't mean the developer will know which buffer that actually is.  Sure, it helps distinguish two calls to the same function that both use buffers, but not really much more than that.

If we really wanted to do that we'd probably need to keep track of the spot where the buffer is created (e.g. have the UI indicate the identifier of the buffer next to the command that created it, something like `createBuffer() -> (@1)`).  For that, you'll probably want to either adjust the IDL generator to capture the return value from things that are instrumented using a `CallTracer` or add specific `InspectorInstrumentation` hooks in the functions that create these objects (e.g. `WebGLRenderingContextBase::createBuffer`).

As far as having a `HashMap`, you'd probably want that to be on `InspectorCanvas` alongside some of the other things we already have there (e.g. `DuplicateDataVariant`).

grorg replied:
Yeah, we definitely need to have `createBuffer` tell the user what the identifier is, but I think that can come later. For now it's still pretty helpful to know that parameters are different.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/RecordingActionTreeElement.js:
```diff
@@ -101,6 +105,7 @@ WI.RecordingActionTreeElement = class RecordingActionTreeElement extends WI.Gene
             case WI.Recording.Swizzle.WebGLVertexArrayObject:
                 parameterElement.classList.add("swizzled");
                 parameterElement.textContent = WI.Recording.displayNameForSwizzleType(swizzleType);
+                parameterElement.textContent += ": " + parameter;
```
We should check to make sure that there is a `parameter` value before attempting to use it, as using a newer Web Inspector with an older target (e.g. an iPhone running an older iOS) or a WebGL recording created with an older Web Inspector will have problems with this.
```
parameterElement.textContent = WI.Recording.displayNameForSwizzleType(swizzleType) + (parameter ? (": " + parameter) : "");
```

dcrousso replied:
based on my above comment maybe we should have it be
```
parameterElement.textContent = WI.Recording.displayNameForSwizzleType(swizzleType) + (parameter ? (`: (@${parameter})`) : "");
```

djg replied:
Before this change, all WebGL objects are recorded with a parameter value of `0`. I'll check if OpenGL `id` is a valid value or not. I suspect not because `0` is used to unbind and object from an attachment point.

djg replied:
> based on my above comment maybe we should have it be
> 
> ```
> parameterElement.textContent = WI.Recording.displayNameForSwizzleType(swizzleType) + (parameter ? (`: (@${parameter})`) : "");
> ```

I've implemented this and pushed.

dcrousso replied:
something I forgot to mention, please make sure that this extra text is not included if the user copies the selected action (see the `copyText` variable inside the `WI.RecordingActionTreeElement.prototype._generateDOM`)

dcrousso reviewed (changes requested):
please see my comment about it not including the `(@...)` in the copy text

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/RecordingActionTreeElement.js:
```diff
@@ -112,6 +116,28 @@ WI.RecordingActionTreeElement = class RecordingActionTreeElement extends WI.Gene
         }
 
+        function copyTextForElement(element, swizzleType) {
+            const textContext = element.textContent;
```
Style: We only use `const` for things that remain constant between Web Inspector sessions (i.e. actual constants).  Please use `let` instead.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/RecordingActionTreeElement.js:
```diff
@@ -112,6 +116,28 @@ WI.RecordingActionTreeElement = class RecordingActionTreeElement extends WI.Gene
+            case WI.Recording.Swizzle.WebGLTransformFeedback:
+            case WI.Recording.Swizzle.WebGLVertexArrayObject:
+                // Strip-off trailing object id beginning at ':' to the end of the string
+                return textContent.replace(/:.*$/, "");
```
rather than manipulate the string afterwards, i think it might be cleaner to have `createParameterElement` return a `{parameterElement, parameterCopyText}`
```
function createParameterUI(parameter, swizzleType, index) {
    let parameterCopyText = "";

    ...
    case WI.Recording.Swizzle.WebGLVertexArrayObject:
        parameterCopyText = WI.Recording.displayNameForSwizzleType(swizzleType);

        parameterElement.classList.add("swizzled");
        parameterElement.textContent = parameterCopyText + (parameter ? (`: (@${parameter})`) : "");
        break;
    ...

    if (!parameterCopyText.length)
        parameterCopyText = parameterElement.textContent;

    return {parameterElement, parameterCopyText};
}
```

and then use it like so
```
...
let {parameterElement, parameterCopyText} = createParameterElement(parameter, swizzleType, i);
parametersContainer.appendChild(parameterElement);

if (i)
    copyText += ", ";

copyText += parameterCopyText;
...
```

djg committed:
[WebInspector] Display OpenGL object ids in canvas inspector
https://bugs.webkit.org/show_bug.cgi?id=246186
rdar://problem/100875691
Reviewed by Devin Rousso and Dean Jackson.

Record the underlying OpenGL object id associated with each WebGL object and
display these ids along with the WebGL object type in WebInspector canvas
inspector mode to aid determining which objects are being referred to in the
call trace.

* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl-full-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl-snapshots-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl2-full-expected.txt:
* LayoutTests/platform/mac-wk1/inspector/canvas/recording-webgl2-snapshots-expected.txt:
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::processArgument):
* Source/WebInspectorUI/UserInterface/Views/RecordingActionTreeElement.js:
(WI.RecordingActionTreeElement._generateDOM.createParameterElement):
(WI.RecordingActionTreeElement._generateDOM):

Canonical link: https://commits.webkit.org/255565@main
